EIPS/eip-4750.md

```diff
@@ -103,6 +99,8 @@ In addition to container format validation rules above, we extend code section v
 3. `RJUMP`, `RJUMPI` and `RJUMPV` immediate argument value (jump destination relative offset) validation:
     1. Code section is invalid in case offset points to a position outside of section bounds.
     2. Code section is invalid in case offset points to one of two bytes directly following `CALLF` instruction.
+4. First code section is invalid if it contains a `RETF` instruction.
```

Comment:
This is superseded by non-returning functions in 6206, but those aren't a thing in 4750 yet, so this rule needs to be added for completeness

Reply:
Stack validation also is defined in a later EIP, yet we're speccing this out having it in mind. So I think it's fine to refer to unreachableness, I added a suggestion about to `RETF` section. This item here I would probably remove.

(it's also not complete, code section is also invalid if it contains `JUMPF` to returning)

Reply:
> (it's also not complete, code section is also invalid if it contains JUMPF to returning)

But JUMPF in particular is not yet known, and is not required for this spec (also including 5450, as linked and mentioned) to be complete.

I think the point (4.) makes the spec self-contained and independent of at least one EIP (6206). One can also argue that it makes 4750 compatible with 6206 - the rule in 6206 is a generalization of this (4.) one, and by that, 6206 doesn't invalidate code which would be otherwise valid without 6206.

I would prefer to strive to decouple the EIPs wherever possible, like here we're making 6206 an optional extension to 4750. 5450 is much harder to leave out from this EIP than 6206

Reply:
@gumb0 I pushed a version of the document still decoupled from 6206, as I would suggest to proceed, but have a version with 6206 referenced (w/ JUMPF and non-returning fs, more in line with your suggestion) stashed away.

PTAL, I would opt to remain decoupled from 6206 as long as reasonably possible.

Reply:
I think trying to make EIPs independent of each other is a lost cause by now. 

My concern is potential confusion for the people reading two EIPs, seeing these two different versions of the validation rule and needing to go through mental effort to figure out how these versions relate to each other and what supersedes what. It's not trivial to figure out, it's formulated very differently in 6206.

Reply:
> I think trying to make EIPs independent of each other is a lost cause by now.

For 3540, after feedback received, we strived to remove as many dependencies as possible, see [here](https://github.com/ethereum/EIPs/pull/8152#pullrequestreview-1922905668) and [here](https://github.com/ethereum/EIPs/pull/8152/commits/b45d6076c2e9bd59b211eb9e4d1e0d85eff089ed).

I think we're giving up too early on this 4750->6206 reference, and should also make the effort to decouple. That's also the main reason for the EIPs - to spec out new concepts in isolation, without needing to jump around different docs.

> My concern is potential confusion for the people reading two EIPs, seeing these two different versions of the validation rule and needing to go through mental effort to figure out how these versions relate to each other and what supersedes what. It's not trivial to figure out, it's formulated very differently in 6206.

This is a fair point, there should be no confusion, but I think it is not that bad, as the 4750 rules are generalized (not superseded really, I used a wrong word) and made implicit by 6206 rules. I reread 6206 and it seems fine to me.

I pushed the "6206-ish" version in a commit [a12f35ec](https://github.com/pdobacz/EIPs/commit/a12f35ec0f900c33748f8cbbbad199f7e16a67d5) for comparison. The diff is really small and shows the 6206 rules generalize 4750 rules nicely. And if there still remains risk of confusion, maybe better to clarify in 6206 ("rule x generalizes rule y from 4750" or just generally "Some rules generalize rules introduced in 4750")